src/java_bytecode/java_object_factory.cpp

```diff
@@ -110,42 +257,155 @@ void gen_nondet_init(
+  std::string binary_width(java_int_width - significant_bits.length(), '0');
+  return constant_exprt(binary_width += significant_bits, type);
+}
+
+void gen_nondet_state::gen_nondet_array_init(const exprt &expr, const source_locationt &loc)
+{
+  assert(expr.type().id()==ID_pointer);
+  const typet &type=ns.follow(expr.type().subtype());
+  const struct_typet &struct_type=to_struct_type(type);
+  assert(expr.type().subtype().id() == ID_symbol);
+  const typet &element_type=static_cast<const typet &>(expr.type().subtype().find(ID_C_element_type));
+
+  auto max_length_expr=as_number(max_nondet_array_length,java_int_type());
+
+  typet allocate_type;
+  symbolt &length_sym=new_tmp_symbol(symbol_table,"nondet_array_length");
+  length_sym.type=java_int_type();
+  const auto &length_sym_expr=length_sym.symbol_expr();
+
+  // Initialise array with some undetermined length:
+  gen_nondet_init(length_sym_expr,false,irep_idt(),loc,false,false);
+
+  // Insert assumptions to bound its length:
+  binary_relation_exprt assume1(length_sym_expr,ID_ge,
+                                as_number(0, java_int_type()));
+  binary_relation_exprt assume2(length_sym_expr,ID_le,
+                                max_length_expr);
+  code_assumet assume_inst1(assume1);
+  code_assumet assume_inst2(assume2);
+  init_code.move_to_operands(assume_inst1);
+  init_code.move_to_operands(assume_inst2);
+
+  side_effect_exprt java_new_array(ID_java_new_array,expr.type());
+  java_new_array.copy_to_operands(length_sym_expr);
+  java_new_array.set("skip_initialise",true);
+  java_new_array.type().subtype().set(ID_C_element_type,element_type);
+  codet assign = code_assignt(expr,java_new_array);
+  assign.add_source_location()=loc;
+  init_code.copy_to_operands(assign);
+
+  exprt init_array_expr=member_exprt(dereference_exprt(expr, expr.type().subtype()),
+                                     "data", struct_type.components()[2].type());
+  if(init_array_expr.type()!=pointer_typet(element_type))
+    init_array_expr=typecast_exprt(init_array_expr,pointer_typet(element_type));
+
+  // Interpose a new symbol, as the goto-symex stage can't handle array indexing via a cast.
+  symbolt &array_init_symbol=new_tmp_symbol(symbol_table,"array_data_init");
+  array_init_symbol.type=init_array_expr.type();
+  const auto &array_init_symexpr=array_init_symbol.symbol_expr();
+  codet data_assign = code_assignt(array_init_symexpr,init_array_expr);
+  data_assign.add_source_location() = loc;
+  init_code.copy_to_operands(data_assign);
+
+  // Emit init loop for(array_init_iter=0; array_init_iter!=array.length; ++array_init_iter)
+  //                  init(array[array_init_iter]);
+  symbolt &counter=new_tmp_symbol(symbol_table,"array_init_iter");
+  counter.type=length_sym_expr.type();
+  exprt counter_expr=counter.symbol_expr();
+
+  init_code.copy_to_operands(
+    code_assignt(counter_expr,as_number(0, java_int_type())));
+
+  std::string head_name=as_string(counter.base_name)+"_header";
```

Comment:
the variable name is generated using `new_tmp_symbol` which gives unique / non-Java identifiers